### ISSUE TITLE
IX - Manual selection of training set

### ISSUE THREAD

**Issue by txau** [huridocs/uwazidocs#8408]
Users need some control over the samples used for training. An MVP of this feature does:
- Add a column to the table with a checkbox to select/unselect as part of the training set
- The table is sortable, so users can quickly locate the training samples
- The "Train" button presents options to train with the manually selected training set or all other valid samples

This issue solves these other: https://github.com/huridocs/uwazi/issues/8166 https://github.com/huridocs/ml-backlog/issues/134 https://github.com/huridocs/uwazi/issues/8281 https://github.com/huridocs/uwazi/issues/8166

In order to properly handle this in the service, the easiest approach would be sending the "use for training" property of each suggestion as part of the training data. Let the service decide how to properly handle these samples.

**Comment by RafaPolit** [comment]
The data sent to the service, if kept as is described above, will have low chances of sending "use for training" samples within the limited samples sent.  So I actually think that we should do this at the backend level and pick ourselves the "use for training" samples with the following criteria:
- The user actually trains with 3 possible options: a. train picking randomly, b. use only use-for-training samples, c. use use-for-training first and fill quota with random data.
- we then either: do as we have, send only selected samples for training (truncating if there are more than what the service expects), or pick samples marked for training and complete them with the missing samples.

I don't see in this process where would the external service actually come to play a part in decision making.  If we opt for that route, we would be uploading materials to the service (which happens first) that could end up being discarded.  Maybe I'm missing something here?

**Comment by txau** [comment]
The service needs to know if the user marked a sample as "use for training" so it can use it for training despite the value being empty. Currently the service ignores empty values for training.

**Comment by RafaPolit** [comment]
I see. Ok. Yeah, that we need to check on the backend process as well, we could be filtering out empty values.  We need to avoid that for things marked specifically as "use for training".

**Comment by Zasa-san** [comment]
For the front end the remaining work is:
- Check the sidepanel filters 

In the train modal add the option to select:
- Only marked for training
- Marked for training + labeled data 
- Update the route to take the new parameter:
`samplePolicy: 'only_marked' | 'marked_plus_labeled'`

On the table we need to signal suggestions as "Use for training". The new endpoint is

```
/api/suggestions/training-set

type Request = {
    extractorId: string;
    suggestionIds: string[];
    useForTraining?: boolean | undefined;
}
```
It can mark one o several suggestions to be used for training.

This should be done in https://github.com/huridocs/uwazi/pull/8461

**Comment by Zasa-san** [comment]
For the UI update of the Train modal were users can select what sample to use for training will will add the following options:
```
Training sample: ⓘ                                       
 ○ Marked for training only                             
 ○ Marked for training + all labeled entries 
```

Tooltip:
```
Choose which data to use for training:
• Marked for training only: Uses only entries you marked as 
  "Use for training", including those with empty values
• Marked for training + all labeled entries: Includes entries 
  marked for training plus any other entries that have values
```

**Comment by Zasa-san** [comment]
This is a showcase of how suggestions will be marked as part of the training set

<img width="1636" height="202" alt="Image" src="https://github.com/user-attachments/assets/1f23cf5c-347f-42f4-84bb-fa0a0176edce" />

Check for used, X for not used. The users can toggle that.

**Comment by aphilop** [comment]
@Zasa-san 
I am afraid that using these icons as interactive elements might be confusing to users. We use the same check symbol for accepted suggestions and there the icon is not interactive (inconsistency).
Why not use the toggle button we already use somewhere else? 

<img width="53" height="34" alt="Image" src="https://github.com/user-attachments/assets/906887b9-c9ad-4cef-9c66-4878eb3f0b0a" />

We could even enhance it to have the check/X in the round part (?)

**Comment by Zasa-san** [comment]
> Why not use the toggle button we already use somewhere else? 

Because that toggle button animates. Since we are using it to trigger a server request, and it's on a table that will re-render completely because of the request, it's complicated to prevent flickering on the button as it changes due to user interaction combined with the revalidation of data from the server.

I think that users hovering over the current implementation of the button and having the cursor change to reflect that there's an action there is indicator enough.

Can we go with the current solution at least for now so that we don't delay the functionality solely for the visuals? We could consider having a toggle slider on the column as a nice to have perhaps?

**Comment by aphilop** [comment]
> Can we go with the current solution at least for now so that we don't delay the functionality solely for the visuals? We could consider having a toggle slider on the column as a nice to have perhaps?

Yes, ok. Let's do that.

**Comment by txau** [comment]
- I think we could find a better color code/icon for the use for training column. The red icon with an X looks like there is a problem with the suggestion.
- There is a problem with the UX flow, when I open the side panel for a suggestion and set the right value, I normally would want to use it for training. It is difficult to locate it back in the list or, if you have filters, it may get lost if it does not match your current filter. Maybe it makes sense of manually inputted values to automatically put them as use for training?
- I have a issue with a title extractor: https://right.staging.uwazi.io/en/settings/metadata_extraction/suggestions/68e2cb8273cc928707031fcc the servoce only tests one method and gives 100% accuracy (so it wont try other methods) but all suggestions are wrong (no contezt/blank). Are we sending the correct data to the service?
- Another extractor also misbehaves, https://right.staging.uwazi.io/en/settings/metadata_extraction/suggestions/68cd9eb07221a6d929820134 processed suggestions all come as match and mismatch always (check screeshot).

<img width="732" height="551" alt="Image" src="https://github.com/user-attachments/assets/f5ccf516-f537-4a53-85de-77e98cce0190" />

**Comment by aphilop** [comment]
After an initial training, I tried to "Process extractor" asking suggestions for all remaining entries:

<img width="1442" height="1338" alt="Image" src="https://github.com/user-attachments/assets/8c31ea16-f3d8-4fb6-a4e1-20d0dfe7e4cd" />


The modal dialog remained visible with the button deactivated and then after some time I got a timeout error.
Then I tried to "Process extractor" with the default 1000 value for getting suggestions and got another error (unfortunately did not capture it, but was something like "extractor with ID xxx is not ready"). Then after a while, the system indicated that is finding suggestions.

<img width="1431" height="1249" alt="Image" src="https://github.com/user-attachments/assets/11b4fe99-032a-461e-a1ba-d7a114448f54" />

**Comment by aphilop** [comment]
UX problem with the "Train model" modal dialog (observed with Safari):
1. Open the dialog
2. Select "Marked for training + all labeled entries"

<img width="559" height="202" alt="Image" src="https://github.com/user-attachments/assets/de1ee5a2-1332-48df-95a1-751dccaa6057" />

3. Check "Find suggestions after training"

<img width="553" height="202" alt="Image" src="https://github.com/user-attachments/assets/be7a00c2-4709-41f7-a97c-eabf8b4b9031" />

4. The system selects "Marked for training only" automatically, thus the user has to reselect the previously selected option.

**Comment by Zasa-san** [comment]
@aphilop 

> The system selects "Marked for training only" automatically, thus the user has to reselect the previously selected option.

To be fixed in https://github.com/huridocs/uwazi/pull/8487

> The modal dialog remained visible with the button deactivated and then after some time I got a timeout error.

This sounds like the response from the server took a long time, that's why the modal continued open with the button disabled, it was 'submitting'

@txau 

Would this be better for icons?

<img width="1670" height="722" alt="Image" src="https://github.com/user-attachments/assets/e84407a5-c157-49a0-ac6b-aa09293763ab" />

Should we automatically mark the suggestion to be used for training when the user saves it from the sidepanel?

**Comment by txau** [comment]
@Zasa-san the problem I see with the icon is lacking color code, can we do the check mark for the selected ones green?

Yes, I think manually edited should be automatically marked as "good data" (select for training)

**Comment by Zasa-san** [comment]
@RafaPolit @txau 

- When saving an entity via the sidepanel we will also trigger marking it as used for training.
-  ~If the user accepts a suggestion then we mark it as used for training too (except when auto-accepting).~ (Let's table this for now since accepting a suggestion is no longer an immediate action, and the client is notified via sockets about the acceptance status for all accepted suggestions).

**Comment by RafaPolit** [comment]
> Yes, I think manually edited should be automatically marked as "good data" (select for training)

I don't think this is a good idea.  What happens if the user is editing the value of a "unusable for train" scenario where they are simply putting in the value... all of sudden we are marking that as "use for training" for the user, without them probably even noticing?  If they want this, this is a click away... why add a "hidden" action to a thing they did not intend to do explicitly?

cc @txau

**Comment by txau** [comment]
I get it @RafaPolit but I want to wrap this up already. This option is better than nothing imo

**Comment by aphilop** [comment]
I agree with @RafaPolit. Hidden functions that are not transparent to the users are not a good UX practice. 
If at all, this shouldn't be done automatically but offered as an option within the side panel.

**Comment by txau** [comment]
Within the side panel: when the user hits "save" the panel closes automatically, meaning that the users would need to remember to hit "Use for training" before "Save", or change the behavior so it does not close automatically on "Save" so we add extra clicks. 

As stated in the comments, if you have a filter such as "mismatches", upon save, the suggestion does no longer belong to that filter and disappears from the list. So the user has to find it later and hit the use for training. Not great, but not too bad.

On the other hand, if the suggestion that has been manually filled in the side panel is automatically added to "use for training", you are done quickly. In the case of the suggestion not being a good fit for the "use for training" set, it is way easier to filter by "Use for training" and remove the unfit ones, since it is going to be a smaller set on most cases.

A potential in between. We add a check box besides the "Save" button that is active by default so users can unselect it right before clicking "Save"

The point is that I was trying to cut corners and be done with this.

**Comment by Zasa-san** [comment]
Proof of concept. Maybe we can create a new issue to implement choosing to add to training set on entity save

<img width="1920" height="895" alt="Image" src="https://github.com/user-attachments/assets/20adca68-7cf0-41ea-b408-c58ee1fb83b4" />

**Comment by aphilop** [comment]
I like this solution as it provides transparency of what would happen and gives control back to the user.
I would suggest to use the label: "Use for training"  (instead of "Add to training set") to match what we already have in the table UI.

**Comment by aphilop** [comment]
Suggested improvement extracted in #8502 for further handling
